File: python/src/ev2/ut5_pooexcepcionesio/u03ficherosio/teoria/U03_LeerTexto.py
"""
U03 — Leer texto: read() y split por lineas.

OBJETIVO:
  - Leer todo el fichero con read() en una cadena llamada contenido.
  - Observar los saltos de linea '\\n' dentro de la cadena.
  - Trocear con split("\\n") para obtener una lista de lineas (personas).

Autor: Agustin. A. Marquez. Pina
Contacto: agu1406@outlook.es
Repositorio GitHub: https://github.com/Agu1406/ClasesParticulares
Sitio web: https://www.agustinmarquez.dev
"""

# Creamos datos.txt al inicio (asi el ejemplo funciona solo).
fichero = open("datos.txt", "w", encoding="utf-8")
fichero.write(
    "Ana,21,Madrid,Ingeniería,Leer\n"
    "Luis,19,Valencia,Medicina,Fútbol\n"
    "Carmen,22,Sevilla,Historia,Pintar\n"
    "Pedro,20,Bilbao,Economía,Ajedrez\n"
    "Lucía,23,Barcelona,Biología,Viajar\n"
)
fichero.close()

print("\n¡INICIO DEL PROGRAMA!\n")
"""
PRIMERA PARTE — read() guarda TODO en una sola cadena.
"""
print("¡DEMO 1: contenido = fichero.read()!\n")

fichero = open("datos.txt", "r", encoding="utf-8")
contenido = fichero.read()
fichero.close()

print(contenido)
print("---")
print("La variable contenido es un str. Entre lineas hay el caracter \\n.")
print(repr(contenido[:40]), "...")

"""
SEGUNDA PARTE — split("\\n") convierte la cadena en lista de lineas.
"""
print("\n¡DEMO 2: personas = contenido.split('\\n')!\n")

personas = contenido.split("\n")
# Si el fichero acaba en \\n, split deja un ultimo elemento vacio: lo quitamos.
personas = [p for p in personas if p != ""]
print(personas)
print(f"Numero de personas (lineas): {len(personas)}")

print("\n¡FIN DEL PROGRAMA!\n")

File: csharp/src/ev2/ut5_pooexcepcionesio/u03ficherosio/teoria/U03_LeerTexto.cs
/*
U03 — Leer texto de ficheros.

File.ReadAllText devuelve todo el archivo como un unico string.
File.ReadAllLines devuelve un array de strings, una entrada por linea.
Es util recorrer lineas con foreach cuando procesas registro a registro.

OBJETIVO:
  - Preparar un fichero de ejemplo con WriteAllText.
  - Leer todo de golpe con ReadAllText.
  - Leer linea a linea con ReadAllLines y foreach { }.
  - Contar lineas y mostrar cada una numerada.

Autor: Agustin. A. Marquez. Pina
Contacto: agu1406@outlook.es
Repositorio GitHub: https://github.com/Agu1406/ClasesParticulares
Sitio web: https://www.agustinmarquez.dev
*/

public class Program
{
    static void ImprimirMenu()
    {
        Console.WriteLine("=== U03 Leer texto de ficheros ===");
        Console.WriteLine("1. Preparar fichero");
        Console.WriteLine("2. ReadAllText");
        Console.WriteLine("3. ReadAllLines + foreach");
        Console.WriteLine("4. Lectura segura con Exists");
        Console.WriteLine("0. Salir");
    }

    static void Main()
    {
        int opcion;
        do
        {
            ImprimirMenu();
            Console.Write("Introduce una opcion -> ");
            opcion = int.Parse(Console.ReadLine()!);
            Console.WriteLine();

            switch (opcion)
            {
                case 1:
                    DemoPrepararFichero();
                    break;
                case 2:
                    DemoReadAllText();
                    break;
                case 3:
                    DemoReadAllLines();
                    break;
                case 4:
                    DemoLecturaSegura();
                    break;
                case 0:
                    Console.WriteLine("Saliendo...");
                    break;
                default:
                    Console.WriteLine("Opcion no valida.");
                    break;
            }

            Console.WriteLine();
        } while (opcion != 0);
    }

    /*
    PRIMERA PARTE — Crear contenido de prueba con varias lineas.
    */
    static void DemoPrepararFichero()
    {
        Console.WriteLine("¡DEMO — Preparar fichero de lectura!\n");

        string fichero = "demo_ut5.txt";
        File.WriteAllText(fichero, "Linea A: inicio\nLinea B: medio\nLinea C: final\n");
        Console.WriteLine("Fichero demo_ut5.txt preparado con 3 lineas.");
    }

    /*
    SEGUNDA PARTE — ReadAllText: un solo string con saltos \n incluidos.
    */
    static void DemoReadAllText()
    {
        Console.WriteLine("¡DEMO — ReadAllText (todo el texto)!\n");

        string fichero = "demo_ut5.txt";
        string todoElTexto = File.ReadAllText(fichero);
        Console.WriteLine("--- Contenido completo ---");
        Console.WriteLine(todoElTexto);
        Console.WriteLine($"Caracteres totales: {todoElTexto.Length}");
    }

    /*
    TERCERA PARTE — ReadAllLines + foreach: procesar linea por linea.
    */
    static void DemoReadAllLines()
    {
        Console.WriteLine("¡DEMO — ReadAllLines + foreach!\n");

        string fichero = "demo_ut5.txt";
        string[] lineas = File.ReadAllLines(fichero);
        Console.WriteLine($"Numero de lineas: {lineas.Length}");
        Console.WriteLine("--- Recorrido con foreach ---");

        int numero = 1;
        foreach (string linea in lineas)
        {
            Console.WriteLine($"  {numero}. {linea}");
            numero++;
        }
    }

    /*
    CUARTA PARTE — Comprobar existencia antes de leer (buena practica).
    */
    static void DemoLecturaSegura()
    {
        Console.WriteLine("¡DEMO — Lectura segura con File.Exists!\n");

        string otroFichero = "demo_ut5_inexistente.txt";

        if (File.Exists(otroFichero))
        {
            string contenido = File.ReadAllText(otroFichero);
            Console.WriteLine(contenido);
        }
        else
        {
            Console.WriteLine($"{otroFichero} no existe; no se intenta leer.");
        }
    }
}
